feat: API for handling locale change during page transitions

pmrotule opened the pull request:
Closes #150 

The issue is happening since `nuxt-i18n` is setting the locale from `onNavigate` being called in a middleware which won't wait for the page transition to fade out. Unfortunately, this cannot be fully solved from this module since it is not possible to set a global `pageTransition.beforeEnter` while making sure it won't be overwritten from a `transition.beforeEnter` defined in a page component.

That being said, the new option `skipSettingLocaleOnNavigate` would allow us to workaround the issue. If it is `true`, the locale will not be set on navigate which would let us setting it ourselves from a `beforeEnter` transition hook:

```js
// nuxt.config.js

export default {
  // ...
  plugins: ['~/plugins/router'],

  i18n: {
    // ...
    skipSettingLocaleOnNavigate: true,
  },
}
```

```js
// plugins/router.js

export default ({ app }) => {
  app.nuxt.defaultTransition.beforeEnter = () => {
    app.i18n.setPendingLocale()
  }
  // Optional: wait for locale before scrolling for a smoother transition
  app.router.options.scrollBehavior = async (to, from, savedPosition) => {
    // Make sure the route has changed otherwise `beforeEnter` will never be called
    if (to.name !== from.name) await app.i18n.waitForPendingLocale()
    return savedPosition || { x: 0, y: 0 }
  }
}
```

```js
// in any page component

export default {
  transition: {
    beforeEnter() {
      this.$i18n.setPendingLocale()
    },
  }
}
```

I'll wait for your feedback before adding test and documentation. I hope it makes sense.

pmrotule committed:
Add option to skip setting locale on navigate

Useful if you want to wait for the page transition to enter before setting the locale

pmrotule committed:
Add helper to wait for pending locale

rchl commented on src/templates/plugin.main.js:
```diff
@@ -221,6 +242,8 @@ export default async (context) => {
+    i18n.setPendingLocale = setPendingLocale
+    i18n.waitForPendingLocale = waitForPendingLocale
     i18n.__baseUrl = app.i18n.__baseUrl
   }
```
Should also copy those internal properties that you've added (`__pendingLocale` and other).

rchl commented on src/templates/plugin.main.js:
```diff
@@ -174,11 +175,31 @@ export default async (context) => {
+  }
+
+  const waitForPendingLocale = async () => {
+    return !app.i18n.__pendingLocale || (await app.i18n.__pendingLocalePromise)
```
Would prefer this as it would be more correct from the perspective of the types (I'm planning to enable more strict type-checking at some point):
```suggestion
    if (app.i18n.__pendingLocale) {
      await app.i18n.__pendingLocalePromise
    }
```

rchl reviewed (changes requested):
It's not the prettiest solution but I don't know better (better would probably require some changes in Nuxt).

So I'm fine with it for now.

pmrotule committed:
Address feedback on pending locale

MartinMa commented:
Very interesting solution.

@pmrotule Is there a way to let i18n-module call the `setPendingLocale()` function on its own? You mentioned `beforeEnter` but there can only be one `beforeEnter` listener I suppose? Is `beforeEnter` the best possible time to change the locale? When exactly is it called (it isn't actually specified in the docs)? Right before the new page begins to transition in? There would still be an edge case when the old page transitions out (say slide to the left) and the new page transitions in at the same time (say slide in from the right). In this edge case both translations should be visible at the same time. I mean, maybe we just have to live with the fact that this doesn't work.

pmrotule commented:
> Very interesting solution.
> 
> @pmrotule Is there a way to let i18n-module call the `setPendingLocale()` function on its own? You mentioned `beforeEnter` but there can only be one `beforeEnter` listener I suppose? Is `beforeEnter` the best possible time to change the locale? When exactly is it called (it isn't actually specified in the docs)? Right before the new page begins to transition in? There would still be an edge case when the old page transitions out (say slide to the left) and the new page transitions in at the same time (say slide in from the right). In this edge case both translations should be visible at the same time. I mean, maybe we just have to live with the fact that this doesn't work.

@MartinMa `beforeEnter` is indeed the exact time where you need to change it since it is called by the `<transition>` wrapping up your `<router-view />` which differs from the `router.beforeEach` hook. Basically, when you call `switchLocalePath`, it triggers a navigation which will call up to 3 different hooks at the same time:
- `router.beforeEach` will be called
- Any global or matching middleware as well
- The router-view `<transition>` will start

The problem is, `nuxt-i18n` is switching the locale from a middleware so it's switched right away without waiting for the transition to fade out the current page. Unfortunately, `setPendingLocale()` cannot be called from `nuxt-i18n` because it has to be called when the new page transitions in which is happening on `beforeEnter`, but `beforeEnter` can be set globally as well as from page components and in this case, there is really no way of accessing it. The only way I could see it working (like @rchl quickly mentioned) would be by adding a new config in Nuxt allowing to define a `pageTransition.beforeEnter` that couldn't be overwritten from a page component (something like `beforeAnyRouteEnter`).

By the way, sorry about the delay, I should find time in the next few days to wrap up this PR.

pmrotule committed:
Update documentation for skipSettingLocaleOnNavigate

pmrotule commented:
@rchl I finally addressed your feedback and updated the documentation. I tried to add e2e tests, but it wasn't working for me locally. Let me know how you think we should move on with this fix.

AgustinBanchio commented:
Looking forward to this PR

MartinMa commented:
@pmrotule Thank you very much. Looks well thought out and well documented.

pmrotule commented:
@rchl Any feedback? I just rebased again. Should be good to go.

rchl commented:
I will have a look as soon as possible (hopefully this evening).

rchl commented on docs/content/en/api.md:
```diff
@@ -108,6 +108,22 @@ Instance of [VueI18n class](http://kazupon.github.io/vue-i18n/api/#vuei18n-class
 
   Switches locale of the app to specified locale code. If `useCookie` option is enabled, locale cookie will be updated with new value. If prefixes are enabled (`strategy` other than `no_prefix`), will navigate to new locale's route.
 
+#### setPendingLocale
```
I fear that it's called too similar to `setLocale` and could confuse people. Maybe it should be renamed to something else? Maybe `flushPendingLocale`, or `finalizePendingLocaleChange`?

Then I guess the `waitForPendingLocale` should be changed too, for parity. Maybe `finalizePendingLocaleChange` and `waitForPendingLocaleChange`?

pmrotule replied:
Makes sense. I went for `finalizePendingLocaleChange` and `waitForPendingLocaleChange` as you suggested.
https://github.com/nuxt-community/i18n-module/pull/963/commits/a4c0808d7993ec9686a5cf232695fcaf65973b2f

rchl commented on docs/content/en/lang-switcher.md:
```diff
@@ -98,3 +98,46 @@ export default {
+
+## Wait for page transition
+
+By default, the locale will be changed right away when calling [`switchLocalePath(path)`](./api#switchlocalepath) which means that if you have a page transition, it will fade out the page with the text already switched to the new language and fade back in with the same content.
```
> By default, the locale will be changed right away when calling `switchLocalePath(path)`

This statement feels wrong as `switchLocalePath` doesn't change locale but only generates an URL that can be used to change locale through navigating to it.

I've changed it to this in my commit:

> By default, the locale will be changed right away when navigating to a route with a different locale...

Let me know if you disagree.

pmrotule replied:
I totally agree. Thanks for cleaning that up.

rchl committed:
minor formatting changes + added types

rchl reviewed (commented):
Just some small changes and comments.

pmrotule committed:
Rename pending locale methods to avoid confusion

rchl committed:
Minor tweaks

rchl commented:
Great work! Thanks.

Venegrad commented:
7.3 not working
